docs/architecture/adr-104-out-of-band-state-sync.md

```diff
@@ -0,0 +1,215 @@
+led to this ADR. At first glance, their proposal might look identical to the
+solution discussed here. The main distinction is that their proposal does not
+involve installing application snapshots, but relies on node operators to
+manually synchronize the application state.
+
+More precisely, their proposal is to
+"_allow applications to start with a bootstrapped state_
+(i.e., the application has already a state before the node is started)
+_alongside an empty CometBFT instance_
+(i.e., the node's block store is empty and the state store is at genesis state)
+_using a subsystem of the state sync protocol_".
+
+Starting from an "empty CometBFT instance" is a requirement for running State
+sync, which is maintained in the current proposal.
+The mentioned "subsystem of the state sync protocol" is the light client,
+responsible for verifying the application state and for bootstrapping
+CometBFT's block store and state store accordingly.
+
+The distinction, therefore, lies in the way which the state of the application
+from a running node is transferred to the new node to be bootstrapped.
+Instead of relying on application snapshots, produced by the application but
+handled by State sync, [ADR 083][adr083]/[ADR 103][adr103] assumes that node
+operators are able to manually synchronize the application state from a running
+node (it might be necessary to stop it) to a not-yet-started fresh node.
+
+The main limitation of this approach is that it relies on the ability of node
+operators to properly synchronize the application state between two nodes.
+While experienced node operators are likely able to perform this operation in a
+proper way, we have to consider a broader set of users and emphasize that it is
+an operation susceptible to errors.
+Furthermore, it is an operation that is, by definition, application-specific:
+applications are free to manage their state as they see fit, and this includes
+how and where the state is persisted.
+Node operators would therefore need to know the specifics of each application
+in order to adopt this solution.
+
+## Decision
+
+> This section records the decision that was made.
+> It is best to record as much info as possible from the discussion that happened.
+> This aids in not having to go back to the Pull Request to get the needed information.
+
+State sync should support the bootstrap of new nodes from application snapshots
+obtained out-of-band by node operators.
+
+In other words, the following operation should be, in general terms, possible:
+
+1. When configuring a new node, operators should be able to define where (e.g.,
+   a local file system location) the client side of State sync should look for
+   application snapshots to install or restore;
+1. Operators should be able to instruct the server side of State sync on a
+   running node to export application snapshots, when they are available,
+   to a given location (e.g., at the local file system);
```

Comment:
Is it like a standalone command that'll connect to the p2p network and download the snapshots?

Reply:
No p2p, no network, no download. You instruct the node (e.g., via CLI command) to save a snapshot to disk.

Reply:
Of course, you can save it to a NFS shared disk, so you have network and a form of upload.

Reply:
do you mean export a snapshot from the state running in the local node, not discover and download a snapshot from the network?

Reply:
Discovering and fetching snapshots is what the existing State sync does.

The proposed extension assumes that the node operator will discover and fetch the snapshot in some way and just point to the bootstrapping node (see item 1.) where to find the snapshot data.

Reply:
yeah, but I'd like a standalone that simply download the snapshot from p2p network and save to disk, nothing more, but I guess this could be done on app side.

Reply:
i think all this can be done app side, its simpler if its done app side since there may be gotchas in the app that are hard to account for

Reply:
> i think all this can be done app side

The point here is to create a way to a operator to instruct CometBFT to "export" a snapshot. The snapshot will still be produced by the application, using the same ABCI calls currently in use. The difference is that the snapshot retrieved from the application will not be sent through the network but instead saved to the file system.

Reply:
this is already how state sync works for the sdk. Def adding support for some of the items here makes sense, but we would only need a way to bootstrap the node, like huang yi did. 



> Makes sense to rely on the app-side for fulfilling this requirement! Is there already app-side support for it?

we are merging the prs to make it work. they are all open and approved

Reply:
Links, please?

Reply:
links to?

Reply:
> we are merging the prs to make it work. they are all open and approved

Reply:
> > we are merging the prs to make it work. they are all open and approved
> 
> 

https://github.com/cosmos/cosmos-sdk/pull/16067

https://github.com/cosmos/cosmos-sdk/pull/16061